### PR TITLE
drivers: entropy: add Atmel SAM entropy generator driver

### PR DESCRIPTION
**Pull request by aurel32**
Tested on the Atmel SAM E70 Xplained board.

Signed-off-by: Aurelien Jarno <aurelien@aurel32.net>

### CODE REVIEW

**Comment by subbu147** [comment]
could you please add the TRGN in the supported features section of sam_e70_xplained.rst

**Comment by subbu147** [comment]
Hey, I am just curious, if there is a plan to use interrupts available with TRNG instead of polling .

**Comment by aurel32** [comment]
> Hey, I am just curious, if there is a plan to use interrupts available with TRNG instead of polling .

It's not in my plans at least. There should be a new random number every 336 CPU cycles in the worst case (every 168 CPU cycles in the default clock configuration), so I don't think the polling loop is used a lot. Interrupts will also add some overhead, so I am not sure it is really worth it.

All that said, there might be some corner cases that can benefit from it, and someone might want to implement interrupt support. On my side I prefer to concentrate on USB and flash drivers.

**Review by dbkinder — APPROVED** [review]
+1 for doc changes, thanks.

File: drivers/entropy/CMakeLists.txt [hunk 1]
```diff
@@ -5,5 +5,6 @@ zephyr_library_sources_ifdef(CONFIG_ENTROPY_MCUX_RNGA          entropy_mcux_rnga
 zephyr_library_sources_ifdef(CONFIG_ENTROPY_MCUX_TRNG          entropy_mcux_trng.c)
 zephyr_library_sources_ifdef(CONFIG_ENTROPY_NRF5_RNG           entropy_nrf5.c)
 zephyr_library_sources_ifdef(CONFIG_ENTROPY_STM32_RNG          entropy_stm32.c)
+zephyr_library_sources_ifdef(CONFIG_ENTROPY_SAM_RNG            entropy_sam.c)
```
Comment (MaureenHelm) [inline drivers/entropy/CMakeLists.txt]: Move this up one line to preserve alpha order

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

**Review by galak — CHANGES_REQUESTED** [review]
Missing dt support

**Comment by aurel32** [comment]
> Missing dt support

None of the existing entropy drivers have DT support, they use the HAL or access the CMSIS defined registers. Should I go and define new bindings for entropy drivers? In that case, which name should I use, it seems the Linux kernel use `trng` for the few DTS entries declaring it.

**Comment by galak** [comment]
> None of the existing entropy drivers have DT support, they use the HAL or access the CMSIS defined registers. Should I go and define new bindings for entropy drivers? In that case, which name should I use, it seems the Linux kernel use `trng` for the few DTS entries declaring it.

Yeah, I know, but hopefully this is a pretty trivial dts binding :) [and we'll look at getting the other drivers/soc's to introduce bindings as well].

I would use 'rng' looking at Documentation/devicetree/bindings/rng/.

**Comment by galak** [comment]
I worked up support for Kinetis using DTS, take a look at #11062

**Comment by aurel32** [comment]
@galak, thanks, I'll look at that and will come with a new version.

**Comment by aurel32** [comment]
@galak, I have just pushed a new version which uses DT for the label, the registers and and the peripheral ID.

File: drivers/entropy/CMakeLists.txt (hunk 1 above)
Comment (aurel32) [reply drivers/entropy/CMakeLists.txt]: I have just pushed a new version which keep the entries sorted alphabetically.

**Comment by aurel32** [comment]
Now that PR #11062 has been merged, I have rebased this patch onto master.

**Review by galak — APPROVED** [review]
Looks good to me from a DTS point of view

**Comment by aurel32** [comment]
recheck

**Comment by aurel32** [comment]
Rebased onto master following recent merges

**Comment by aurel32** [comment]
recheck

**Review by MaureenHelm — APPROVED** [review]
@agross-linaro 	Will you please have a look at this?

**Comment by aurel32** [comment]
Rebased following CONFIG_ to DT_ renaming.

**Review by agross-oss — APPROVED** [review]

**Pull request merged**